### ISSUE TITLE
Timer Block (not planned) and pause mechanism

### ISSUE THREAD

**Issue by occoder** [LaboratoireMecaniqueLille/crappy#127]
Hi
It seems to be important to introduce a Timer block into Crappy.
Because it's very often that an experiment needs to be operated according to the time counting, say dwelling at operating point A for X min then dwelling at operating point B for Y min.
Also sometimes, an experiment may need to be paused. Then the remaining time needs to be remembered and resumed later on.
So please consider to introduce a Timer block.
Thanks

**Comment by WeisLeDocto** [comment]
Hi !

> Because it's very often that an experiment needs to be operated according to the time counting

The Generator Block can already put devices on hold following a specific pattern. For example, if it's driving a motor, it can specify times when the motor should stop for a given delay. Using a combination of [Generator](https://crappy.readthedocs.io/en/stable/crappy_docs/blocks.html#crappy.blocks.Generator) and [Modifier](https://crappy.readthedocs.io/en/stable/crappy_docs/modifiers.html#modifiers) objects, it is possible to achieve quite complex command patterns already.

> Also sometimes, an experiment may need to be paused

That is a known current limitation of Crappy, there is no specific mechanism for putting an entire experiment on hold at once. That is because Crappy is mostly designed to achieve reproducibility between several runs: the same code should always drive a given setup in a reproducible way (modulo variability in external factors, e.g. sample stiffness). It would technically be possible to write a Block that drives equipment in an interactive way (through a graphical interface for example). However, combining the interactive approach of such a Block with the deterministic one of the Generator seems a bit contradictory and would likely prove quite difficult.

Another approach could be to act at the Blocks level to temporarily stop their event loop, for example using a global flag. However, a basic, naive implementation wouldn't do: a motor already running won't stop until it is told to do so. Therefore this kind of approach would also require quite some engineering. 

Does that answer your question ? Maybe you had something else in mind for a hypothetical Timer Block ?

**Comment by occoder** [comment]
> It would technically be possible to write a Block that drives equipment in an interactive way (through a graphical interface for example). However, combining the interactive approach of such a Block with the deterministic one of the Generator seems a bit contradictory and would likely prove quite difficult.
> 
Please don't use any interactive element. Leave the UI part to the end user. The user should be reponsible for the conversion of  UI event to trigger signal that can be digested by Crappy.
> Another approach could be to act at the Blocks level to temporarily stop their event loop, for example using a global flag. However, a basic, naive implementation wouldn't do: a motor already running won't stop until it is told to do so. Therefore this kind of approach would also require quite some engineering.
> 
A pause is not a stop.  The pause just freezes every things except for timestamp. If a motor is running in a constant speed, it should keep the speed. If it is accelerating, then pause just does not apply at the accelerating phase. After all, the acceleration won't last forever. Once the motor gets to the stable state, then it can be paused.
> Does that answer your question ? Maybe you had something else in mind for a hypothetical Timer Block ?
The hypothetical Timer Block could be a pure data processing block.
It acts like a gate inbetween a upstream block and a downstream block, i.e. either pass-through or cut-off. 
The key factor is the accurate control (at sub-second level at least) of gating timing.
The time counting task could be a specific thread and its operation is under the control of cmd_label.

**Comment by occoder** [comment]
> Well, the original use case of Crappy is for tensile tests, and in these we definitely don't want the motor to keep running when no data is being acquired. Maybe you have examples in mind when disabling acquisition and control while still letting the actuators running is acceptable/desirable ?
> 
In this specific case of keeping motor speed, it just follows the pause definition, i.e. freezing the current state.
Of course, you can leave the defintion to the end user. E.g. the pause can only be applied to mechanical static states. 

> So, basically, you want to be able to enable or disable a flow of data based on some condition, right ? 
>
No. Disabling of data flow is just one aspect of the pause. BEFORE the time couting, a trigger command could be sent to related blocks to keep all related devices' tates  still. Likewise, AFTER the time couting, another trigger command could be sent as well.

>This is feasible, with variable complexity depending on what the condition is and how it is computed (acquired signal, predefined timestamps, etc.). Now, this Block (more likely Modifier) will only drive the flow of one stream of data.
>
Could you eleberate on how the time counting is done using the "predefined timestamps"? If it is generic enough, can it justify such custom block being designed as a standard Block.

**Comment by occoder** [comment]
> May I ask you what use case you have in mind for such a feature ?
> 
In fact, I'd like to run a burn-in test that involves some conditioning steps. Each conditioning step lasts for certain period.
But at the end of the period, a condition needs to be met. If current step fails to meet the condition, the step will be retried. There is max. time of retry setting that if run out the whole test will be terminated and deemed as a failed test. A standard Timer block may help on time counting for each step and triggering condition check when current period of time runs out.

**Comment by WeisLeDocto** [comment]
I don't see the need for a Block putting the entire experiment on hold here, except maybe if you need to manipulate anything on your setup between conditioning steps. Your use case is clearly quite complex to implement. Still, the following pseudo-code can be a good start:

```python
import crappy

# Generator that outputs the index of the condition to apply
step_path = [{'type': 'Constant', 'value': i, 'condition': f'next_step_id>{i}'} for i in range(...)]
step_id_gen = crappy.blocks.Generator(step_path, cmd_label='step_id'...)

# Perform action on the setup based on the received command
block_1 = SomeBlock(cmd_labels=('value_1', 'value_2', ...), ...)  # Just an example, might be a different syntax
# Acquire data from the setup
block_2 = SomeBlock(...)
...
# These Blocks might actually be a single IOBlock + InOut, or anything else

# User-defined table converting step index to command values
def index_to_values(data);
    if data['step_id'] == 0:
        data['value_1'] = ...
        data['value_2'] = ...
        ...
    elif data['step_id'] == 1:
        data['value_1'] = ...
        data['value_2'] = ...
        ...
    ...
    return data

# Convert step index to actual command values, and pass them to the driver Block
crappy.link(step_id_gen, block_1, modifier=index_to_values)

# Helper class checking whether a step was successful
class DataToIndex(crappy.modifier.Modifier):
    def __init__(self):
        super().__init__()
        self.index = 0
    def __call__(self, data):
        if <condition checking that step was successful>:
            self.index += 1
        data['next_step_id'] = self.index
        return data

# Parse data from setup to check if condition is met, increase next index counter if so 
# to switch to next step, and send to the Generator
crappy.link(block_2, step_id_gen, modifier=DataToIndex())

crappy.start()
```

It might not be sufficient if:

- Each conditioning step follows some different pattern that would be difficult to summarize with a small number of parameters
- There is no clear way for `DataToIndex` to know when a conditioning step is over, and thus to know when to check for the success condition and increment the index counter

However, with the information I have, this is the simplest code I can propose.

**Comment by occoder** [comment]
Thank you @WeisLeDocto , the example is very helpful.
>I don't see the need for a Block putting the entire experiment on hold here, except maybe if you need to manipulate anything on your setup between conditioning steps.
>
Sorry, I forget to mention that during the day time, a worker has to intervene (i.e. to pause the test for some time and then resume afterwards) the setup when he needs to reach the testing area.
>There is no clear way for DataToIndex to know when a conditioning step is over, and thus to know when to check for the success condition and increment the index counter
>
That just where the proposed Time block comes into play, plus the above extra pause/resume functional need during each time counting period.

**Comment by WeisLeDocto** [comment]
> So only at the moment that the step's time runs out, the condition check should execute ONCE.

Can the moment when a step ends be inferred from the acquired data ? If so, something like that would work:

```python

class DataToIndex(crappy.modifier.Modifier):
    def __init__(self, min_delay_between_checks = 100):
        super().__init__()
        self.index = 0
        self.last_t = time.time()
        self.min_delay = min_t_between_checks
    def __call__(self, data):
        if <condition checking that step is over> and time.time() - self.last_t > self.min_delay:
            self.last_t = time.time()
            if <condition checking that step was successful>:
                self.index += 1
        data['next_step_id'] = self.index
        return data
```

Another question, is the intervention of the worker independent of the completion of burn-in steps ? Also, does it happen at times predictable enough to be entered in Crappy before starting a test ? Or should the pauses rather be interactively triggered ?

**Comment by WeisLeDocto** [comment]
In that case, if I understand correctly, you'll need a way to start a pause at any moment. That's what I mean with "interactive trigger", my English is not always on point. This could be achieved using a [Button Block](https://crappy.readthedocs.io/en/stable/crappy_docs/blocks.html#button), that displays a basic UI with a button and increments a counter when clicked. The Timer Block would then pause the experiment for a given delay each time the button is clicked. 

___

To summarize, you have a feature request for a basic "Timer" Block that pauses the Blocks' event loops so that the worker can enter the test area, and in addition to that, it is still a bit unclear how the rest of your script would look like for managing the burn-in steps, right ?

**Comment by occoder** [comment]
> To summarize, you have a feature request for a basic "Timer" Block that pauses the Blocks' event loops so that the worker can enter the test area, 
>
Correct.
>and in addition to that, it is still a bit unclear how the rest of your script would look like for managing the burn-in steps, right ?
>
Each time step has a fixed time of period and relavant condition check at the end of the period. This also needs the Timer block 's help.
As for the workflow that manages the test steps, it seems that Crappy's Generator is not the perfect filt.
I'd like to make a custom output only IOBlock object which uses get_data to feed externally generated command (e.g. through a message queue) to the downstream blocks.

**Comment by WeisLeDocto** [comment]
> This does not have to be implemented as a paused button

What I mean is that a button is already implemented in Crappy in a minimal UI, and could be used for driving a Pause Block.

> What about adding a pause and a resume methods to the Blocks? This does even more make sense, considering the Blocks already are able to response to a global stop notification.

Precisely, they could be able to respond to such a notification given small adjustments in the code. I do see the interest of only pausing actuation Blocks, so that there's no gap in the data acquisition. My only concern is that end users would have to put extra care in the design of their logic/decision Blocks and layout, if these are not paused as well.
Pausing Blocks individually would require some reserved label transmitting only stop and wake up messages. Or, with lower granularity, a per-Block flag indicating whether the Block should respond to a global pause. I have a preference for the latter implementation.

> And it needs to remember the pause moment so that it can resume from where it was paused.

I'm not sure to understand this requirement. If a Block is on hold, its state is frozen and cannot change. It will always resume from where it was paused.

> This also needs the Timer block 's help

To keep the code clean and generic, the "Timer", or rather "Pause" Block, should only be responsible for stopping and resuming other Blocks. It's rather the logic driving the Pause Block that should be shared to whatever object manages the steps. If this object is put on hold when a pause command is issued, it will anyway only operate once the pause ends.

I'm also still failing to understand how pauses in the experiment and step management are related, and why input from the Pause Block would be required to make decision on the next step.

> As for the workflow that manages the test steps, it seems that Crappy's Generator is not the perfect fit.

Yeah I assumed so, given the information you already gave. Can you elaborate on what your requirements are, and how the current implementation fails to meet them ? I'm not fully satisfied with the current Generator implementation, so I'm curious about use cases where it's a limitation.

**Comment by occoder** [comment]
> Precisely, they could be able to respond to such a notification given small adjustments in the code. I do see the interest of only pausing actuation Blocks, so that there's no gap in the data acquisition. My only concern is that end users would have to put extra care in the design of their logic/decision Blocks and layout, if these are not paused as well. Pausing Blocks individually would require some reserved label transmitting only stop and wake up messages. Or, with lower granularity, a per-Block flag indicating whether the Block should respond to a global pause. I have a preference for the latter implementation.
> 
What about treating pause and resume at the same level of stop? This makes the whole concept straight and clear. 

> I'm not sure to understand this requirement. If a Block is on hold, its state is frozen and cannot change. It will always resume from where it was paused.
> 
For example, current step needs to last for 5 min. At the 2nd min, it is paused. Then the step needs to remember there still remains 3 min to go when it is resumed. Of course, there likely are other related states, such as growing counters,  that also need to be stored when paused and restored when resumed.

> To keep the code clean and generic, the "Timer", or rather "Pause" Block, should only be responsible for stopping and resuming other Blocks. It's rather the logic driving the Pause Block that should be shared to whatever object manages the steps. If this object is put on hold when a pause command is issued, it will anyway only operate once the pause ends.
I'm also still failing to understand how pauses in the experiment and step management are related, and why input from the Pause Block would be required to make decision on the next step.
> 
I think the fixed time counting without any pause/resume is the basic function of a Timer block. It'd take the burden of counting time off the test step managers which only sends the parameters of the step including the lasting time to the Timer block. The Timer block also feeds back to the test manager to let it know when it can proceed with the next step.
And pause/resume has nothing to do with making decision on the next step. It's a adavanced function of a Timer block.
> Yeah I assumed so, given the information you already gave. Can you elaborate on what your requirements are, and how the current implementation fails to meet them ? I'm not fully satisfied with the current Generator implementation, so I'm curious about use cases where it's a limitation.
>
The key drawback of the Generator block is that it only support the predefined path. Currently, if the test manager needs to change the path during the runtime, it has to stop the Crappy first before loading the new one. Imagine that what if each test step is dynamically generated (e.g. random test) rather than a predefine set of test steps? Modifier may mend some, but in nature the Generator block had better support more Input features.

**Comment by WeisLeDocto** [comment]
> What about treating pause and resume at the same level of stop?

Do you mean at the same level as `Block.stop()` ? I don't understand what you mean here, and how it translates to an actual design.

> Then the step needs to remember there still remains 3 minutes to go when it is resumed.

I think the simplest option would be to override the `Timer.run()` method so that it counts the time spent paused. The same principle could be implemented in Generators. If the state of the hardware is preserved during a pause, then the actual time spent in that state will be impossible to predict anyway.

> I think the fixed time counting without any pause/resume is the basic function of a Timer block.

I'm not sure that just counting time is worth the extra complexity of having a dedicated Block. It is really just a matter of calling `time.time()` and updating a counter. But this is really an implementation detail.

> And pause/resume has nothing to do with making decision on the next step. It's just an advanced function of a Timer block.

I would still recommend implementing the pause and resume function in a separate Block, to enforce single responsibility principle.

> in nature the Generator block had better support more Input features.

I see, as it is now it can only support a sequential Paths layout. A first improvement could be to support a decision tree instead, and to take decision on the next Path dynamically at runtime. The most challenging aspect would then be to find an elegant and simple way for users to input this decision tree in Crappy.

**Comment by occoder** [comment]
> Do you mean at the same level as `Block.stop()` ? I don't understand what you mean here, and how it translates to an actual design.
> 
Yes,  because you can treat a pause as a special case of a stop and a resume as of a special case of a start.
> I think the simplest option would be to override the `Timer.run()` method so that it counts the time spent paused. The same principle could be implemented in Generators. If the state of the hardware is preserved during a pause, then the actual time spent in that state will be impossible to predict anyway.
> 
This part confuses me that the time span of a pause should be the focus here. Nobody cares how long the pause really take. Instead, caring more about the pause does not mess up the original time setting,  i.e. 5 min = 2min + 3min.
> I'm not sure that just counting time is worth the extra complexity of having a dedicated Block. It is really just a matter of calling `time.time()` and updating a counter. But this is really an implementation detail.
> 
IMHO, time should be treated as first class citizen in Crappy or broadly speaking in any experiment setup. So a standard Timer block is worthwhile.
> I would still recommend implementing the pause and resume function in a separate Block, to enforce single responsibility principle.
> 
Agree. Maybe it's not even necessary to create such a pause Block. Embedding the pause/resume capability to each appropriate block is also an option.
> I see, as it is now it can only support a sequential Paths layout. A first improvement could be to support a decision tree instead, and to take decision on the next Path dynamically at runtime. The most challenging aspect would then be to find an elegant and simple way for users to input this decision tree in Crappy.
>
The Generator is different from other Blocks as its input includes not only from the physical world (generated by the devices) but also from the logic world (generated by the program) and from the human world (generated by the UI). Therefore the interface of Generator's input should be designed to be easy to link to these three worlds. For enough versatility, I'd recommend to design orienting to a populor middleware, such as Redis. This avoids the unnecessary complexity involved.

**Comment by WeisLeDocto** [comment]
> you can treat a pause as a special case of a stop and a resume as of a special case of a start.

I strongly disagree here, initialization and de-initialization should only be performed once during an experiment. The state hardware was left in before the pause could be strongly affected by initialization or de-initalization.

> Nobody cares how long the pause really takes.

That's highly dependent on the context. For instance, if an oven should stay 5min at 1000° but gets paused 5min, the sample inside will stay in total 10min at 1000° and that's what actually matters.

At the very least, a good pause implementation should allow users to decide what happens for hardware when a pause starts. And to decide whether a Timer Block should keep counting during a pausebor not.

> So a standard Timer block is worthwhile.

How would it differ from a Generator using delay=... stop conditions ?

> Therefore the interface of Generator's input is better designed to be easy to link to these three worlds

That's not so much of a concern, any type of input always end up as a key/value pair and is treated similarly by the Generator. The struggle rather is how to make the internal Generator's logic arbitrarily complex while still keeping things clear and manageable for users.

> This would avoid the unnecessary complexity involved

Sounds to me a bit overkill to add external dependency just for solving a very specific and uncommon issue. A super-Generator relying on an external tool should clearly be a complete separate Block.

**Comment by occoder** [comment]
> That's highly dependent on the context. For instance, if an oven should stay 5min at 1000° but gets paused 5min, the sample inside will stay in total 10min at 1000° and that's what actually matters.
> 
It seems that we have very different understanding about what a pause is.
In this specific case, if you can redefine a fixed period of a "pause" state that you need dwell on that that is NOT a pause. It should be deemed as a normal working state, although it appears to be a pause state as not working parameters changing within the current path or test step. 
It's fair to say that only those frozen state that asynchronously happens and its lasting time is not predictable should be called a pause.

**Comment by WeisLeDocto** [comment]
> ONLY those frozen state that asynchronously happens and its lasting time is not predictable should be called a pause

We're on the same line here, I picked 5min as an example, and thought it was obvious that in practice this duration was _a priori_ unknown. My point really was that measuring the duration of a pause can in some cases be relevant, especially if the state of the hardware remains frozen during the pause.

> IMHO, the delay function should have been decoupled from Generator block in the first place.

Interesting, I had never thought about it this way. I agree that decoupling time counting, logic, and signal generation is _in theory_ a good idea. Now, _in practice_, and in the current state of Crappy, a Timer Block would be extremely similar to a Generator with delay conditions. Its arguments would be the same as for a constant Path, and it would also output the same kind of signal. That's why I don't see a need to reinvent the wheel on this specific aspect, the functionality already exists in Crappy.

That being said, I'm actually +1 on a clear, separate management of time / logic / signals in Crappy. I think that's something that should be carefully thought over and will include much larger changes than just removing time counting from the Generator Block. A Timer Block would likely be part of this new architecture.

___

To summarize a bit:

- What I can do on the short-term regarding your specific issue is to implement a pause mechanism at the Block level.
- A Timer Block, although easy to implement, would duplicate a functionality the Generator currently offers. It would be relevant to implement in the context of a complete refactoring of Crappy's signal management, which is unlikely to happen on the short-term.

**Comment by occoder** [comment]
> To summarize a bit:
> 
> * What I can do on the short-term regarding your specific issue is to implement a pause mechanism at the Block level.
> * A Timer Block, although easy to implement, would duplicate a functionality the Generator currently offers. It would be relevant to implement in the context of a complete refactoring of Crappy's signal management, which is unlikely to happen on the short-term.
>
It's great to see that we've landed on a common ground.
Implementing a full-fledged Timer block might be a challenge, especially in a non-realtime OS, like Linux. More difficult, if not impossible, in a Win environment.
So best wishes and may this project prosper.

**Comment by WeisLeDocto** [comment]
Issue closed by #139, that adds a pause mechanism for the Blocks.
I'm keeping the Timer Block feature request in the back of my mind, but I'm not planning it for now.
